### PR TITLE
Introducing a method to the OAuthClientAuthenticator to retrieve the authenticator referred name

### PR DESCRIPTION
**Pull request by ChinthakaJ98**
Ex: PrivateKeyJWTClientAuthenticator referred as private_key_jwt, MutualTLSClientAuthenticator referred as tls_client_auth

## Related Issues
https://github.com/wso2/product-is/issues/16921

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/OAuthClientAuthenticator.java [hunk 1]
```diff
@@ -65,4 +65,11 @@ boolean canAuthenticate(HttpServletRequest request, Map<String, List> bodyParams
      */
     String getClientId(HttpServletRequest request, Map<String, List> bodyParams, OAuthClientAuthnContext
             oAuthClientAuthnContext) throws OAuthClientAuthnException;
+
+    /**
+     * Retrieve the referred name of the authenticator.
+     *
+     * @return      Referred name of the authenticator.
+     */
+    String getReferredName();
```
Comment (Thumimku) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/OAuthClientAuthenticator.java:74]: Can we add it as default method, so it won't affect existing implementations?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/BasicAuthClientAuthenticator.java [hunk 2]
```diff
@@ -269,4 +269,14 @@ protected void setClientCredentialsFromParam(Map<String, List> bodyParams, OAuth
         context.addParameter(OAuth.OAUTH_CLIENT_SECRET, stringContent.get(OAuth.OAUTH_CLIENT_SECRET));
     }
 
+    /**
+     * Retrieve the referred name of the authenticator.
+     *
+     * @return      Referred name of the authenticator.
```
Comment (Thumimku) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/BasicAuthClientAuthenticator.java:275]: I think there is a type?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/PublicClientAuthenticator.java [hunk 3]
```diff
@@ -203,4 +203,14 @@ private void setClientCredentialsFromParam(Map<String, List> params, OAuthClient
...
     }
 
+    /**
+     * Retrieve the referred name of the authenticator.
+     *
+     * @return      Referred name of the authenticator.
+     */
+    @Override
+    public String getReferredName() {
+        return "public_auth";
```
Comment (Thumimku) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/PublicClientAuthenticator.java:213]: Can we make these return as constants for future reference?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/OAuthClientAuthenticator.java [hunk 4]
```diff
@@ -65,4 +65,11 @@ boolean canAuthenticate(HttpServletRequest request, Map<String, List> bodyParams
      */
     String getClientId(HttpServletRequest request, Map<String, List> bodyParams, OAuthClientAuthnContext
             oAuthClientAuthnContext) throws OAuthClientAuthnException;
+
+    /**
+     * Retrieve the referred name of the authenticator.
```
Comment (Thumimku) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/OAuthClientAuthenticator.java:70]: In which context we assigned these names to the authenticators? 
For an example, If I am gonna write new authenticator, what should I follow to derive the referred name?
Can we add usecase of this method?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/PublicClientAuthenticator.java (hunk 3 above)
Comment (ChinthakaJ98) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/PublicClientAuthenticator.java]: Fixed in 684ce335e1241bfce1d6fd56aae4ca08abba580c

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/OAuthClientAuthenticator.java (hunk 1 above)
Comment (ChinthakaJ98) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/OAuthClientAuthenticator.java]: Fixed in 684ce335e1241bfce1d6fd56aae4ca08abba580c

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/OAuthClientAuthenticator.java (hunk 4 above)
Comment (ChinthakaJ98) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/OAuthClientAuthenticator.java:70]: A few sample scenarios are mentioned in the PR description as well. These references can be found in the OpenID spec [1]
[1] https://openid.net/specs/openid-connect-registration-1_0.html#ClientMetadata

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/BasicAuthClientAuthenticator.java (hunk 2 above)
Comment (Thumimku) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/BasicAuthClientAuthenticator.java:276]: Sorry I meant typo.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/OAuthClientAuthenticator.java (hunk 4 above)
Comment (Thumimku) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/OAuthClientAuthenticator.java:70]: So referred name actually token_endpoint_auth_method?

Comment (ChinthakaJ98) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/OAuthClientAuthenticator.java:70]: token_endpoint_auth_method is the attribute that we use in DCR to store the client authentication methods to be used for that application. For that attribute as values we give the referred name. Find below an example.
token_endpoint_auth_method="private_key_jwt"

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/OAuthClientAuthenticator.java [hunk 5]
```diff
@@ -65,4 +65,14 @@ boolean canAuthenticate(HttpServletRequest request, Map<String, List> bodyParams
...
+     *
+     * @return      Referred name of the authenticator.
+     */
+    default String getReferredName() {
```
[the remark this answers is not in the document]
Comment (ChinthakaJ98) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/OAuthClientAuthenticator.java]: Fixed in c7f467e7cc24c313990875d639bb587069ee84a5

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/OAuthClientAuthenticator.java [hunk 6]
```diff
@@ -65,4 +66,14 @@ boolean canAuthenticate(HttpServletRequest request, Map<String, List> bodyParams
...
             oAuthClientAuthnContext) throws OAuthClientAuthnException;
+
+    /**
+     * Retrieve the authentication methods supported by the authenticator.
+     *
+     * @return      Authentication methods supported by the authenticator.
+     */
+    default List<String> getSupportedClientAuthenticationMethods() {
+
+            return new ArrayList<>();
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/OAuthClientAuthenticator.java:77]:
```suggestion
            return Collections.emptyList();
```

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/client/authentication/BasicAuthClientAuthenticatorTest.java [hunk 7]
```diff
@@ -285,6 +286,14 @@ public void testGetClientIdErrorScenario(String headerName, String headerValue,
                 OAuthClientAuthnContext());
     }
 
+    @Test
+    public void testGetSupportedClientAuthenticationMethods() {
+
+        List<String> supportedAuthMethods = basicAuthClientAuthenticator.getSupportedClientAuthenticationMethods();
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/client/authentication/BasicAuthClientAuthenticatorTest.java:292]: Shall we asset the list size is also 2 to make sure additional things are not added.

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/client/authentication/SampleClientAuthenticator.java [hunk 8]
```diff
@@ -74,12 +76,18 @@ public int getPriority() {
...
     @Override
     public boolean isEnabled() {
 
         return this.enabled;
     }
+
+    @Override
+    public List<String> getSupportedClientAuthenticationMethods() {
+
+        return Arrays.asList(SAMPLE_CLIENT_AUTHENTICATOR);
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/client/authentication/SampleClientAuthenticator.java:91]: Lets use a more realistic value here. like client_auth_sample

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/client/authentication/BasicAuthClientAuthenticatorTest.java (hunk 7 above)
Comment (ChinthakaJ98) [reply components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/client/authentication/BasicAuthClientAuthenticatorTest.java:292]: Fixed in 428fe44b8129a99fb54bc22701c96ca5c6dc0bff

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/OAuthClientAuthenticator.java (hunk 6 above)
Comment (ChinthakaJ98) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/OAuthClientAuthenticator.java]: Fixed in 428fe44b8129a99fb54bc22701c96ca5c6dc0bff

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/client/authentication/SampleClientAuthenticator.java (hunk 8 above)
Comment (ChinthakaJ98) [reply components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/client/authentication/SampleClientAuthenticator.java]: Fixed in ef9331d6b71b68ea222d2747bb6d462e626cc762

**Review by janakamarasena — APPROVED** [review]

**Pull request merged**